File: library/lua/eventdispatcher/EventDispatcher.lua
----------------------------------------------------------------------
-- Author:	jaron.ho
-- Date:	2013-11-19
-- Brief:	event dispathcer system
----------------------------------------------------------------------
function CreateEventDispatcher()
	local mEventHandlerMap = {}
	local dispathcer = {}
	-- subscribe event handler: eventId(number,string);func(function);priority(number),1>2>3
	function dispathcer:subscribe(eventId, func, target, priority)
		assert("number" == type(eventId) or "string" == type(eventId), "eventId is not number or string, it's type is "..type(eventId))
		assert("function" == type(func), "func is not function, it's type is "..type(func))
		local handlers = mEventHandlerMap[eventId]
		if nil == handlers then
			handlers = {}
			if "number" ~= type(priority) then
				priority = #handlers
			end
			table.insert(handlers, {func = func, target = target, priority = priority})
			mEventHandlerMap[eventId] = handlers
			return
		end
		for key, handler in pairs(handlers) do
			if func == handler.func then
				return
			end
		end
		if "number" ~= type(priority) then
			priority = #handlers
		end
		table.insert(mEventHandlerMap[eventId], {func = func, target = target, priority = priority})
		table.sort(mEventHandlerMap[eventId], function(a, b) return a.priority < b.priority end)
	end
	-- unsubscribe event: eventId(number,string);func(function)
	function dispathcer:unsubscribe(eventId, func)
		if nil == eventId then
			mEventHandlerMap = {}
			return
		end
		assert("number" == type(eventId) or "string" == type(eventId), "eventId is not number or string, it's type is "..type(eventId))
		local handlers = mEventHandlerMap[eventId]
		if nil == handlers then
			return
		end
		if nil == func then
			mEventHandlerMap[eventId] = nil
			return
		end
		for key, handler in pairs(handlers) do
			if func == handler.func then
				table.remove(mEventHandlerMap[eventId], key)
				break
			end
		end
		if 0 == #mEventHandlerMap[eventId] then
			mEventHandlerMap[eventId] = nil
		end
	end
	-- post event: eventId(number,string);...(any type)
	function dispathcer:post(eventId, ...)
		assert("number" == type(eventId) or "string" == type(eventId), "eventId is not number or string, it's type is "..type(eventId))
		local handlers = mEventHandlerMap[eventId]
		if nil == handlers then
			return
		end
		for key, handler in pairs(handlers) do
			if "function" == type(handler.func) then
				if "table" == type(handler.target) or "userdata" == type(handler.target) then
					handler.func(handler.target, ...)
				else
					handler.func(...)
				end
			end
		end
	end
	return dispathcer
end
----------------------------------------------------------------------
local mEventDispatcher = CreateEventDispatcher()
EventCenter = {}
-- bind event
function EventCenter:bind(eventId, func, target, priority)
	mEventDispatcher:subscribe(eventId, func, target, priority)
end
-- unbind event
function EventCenter:unbind(eventId, func)
	mEventDispatcher:unsubscribe(eventId, func)
end
-- clear event
function EventCenter:clear()
	mEventDispatcher:unsubscribe()
end
-- post event
function EventCenter:post(eventId, ...)
	mEventDispatcher:post(eventId, ...)
end
----------------------------------------------------------------------

File: library/javascript/eventdispatcher/EventDispatcher.js
/***********************************************************************
** Author:	jaron.ho
** Date:	2015-08-06
** Brief:	event dispathcer system
***********************************************************************/
function createEventDispatcher() {
	var dispathcer = {};
    dispathcer._eventHandlerMap = {};
	// subscribe event handler: eventId(number,string);func(function);target(object);priority(number),1>2>3
	dispathcer.subscribe = function (eventId, func, target, priority) {
		if ('string' != typeof(eventId) && 'number' != typeof(eventId)) {
			throw new Error("eventId is not number or string, it's type is " + typeof(eventId));
		}
		if ('function' != typeof(func)) {
			throw new Error("func is not function, it's type is " + typeof(func));
		}
		var handlers = this._eventHandlerMap[eventId];
		if (!(handlers instanceof Array)) {
			handlers = [];
			if ('number' != typeof(priority)) {
				priority = handlers.length;
			}
			handlers.push({func: func, target: target, priority: priority});
			this._eventHandlerMap[eventId] = handlers;
			return true;
		}
		for (var i = 0; i < handlers.length; ++i) {
			if (func == handlers[i].func) {
				return false;
			}
		}
		if ('number' != typeof(priority)) {
			priority = handlers.length;
		}
		this._eventHandlerMap[eventId].push({func: func, target: target, priority: priority});
		this._eventHandlerMap[eventId].sort(function (a, b) {
			return a.priority - b.priority;
		});
		return true;
	};
	// unsubscribe event: eventId(number,string);func(function)
	dispathcer.unsubscribe = function (eventId, func) {
		if ('string' != typeof(eventId) && 'number' != typeof(eventId)) {
			throw new Error("eventId is not number or string, it's type is " + typeof(eventId));
		}
		var handlers = this._eventHandlerMap[eventId];
		if (!(handlers instanceof Array)) {
			return;
		}
		if ('function' != typeof(func)) {
			delete this._eventHandlerMap[eventId];
			return;
		}
		for (var i = 0; i < handlers.length; ++i) {
			if (handlers[i] && 'object' == typeof(handlers[i]) && func == handlers[i].func) {
				handlers.splice(i, 1);
				break;
			}
		}
		if (0 == handlers.length) {
			delete this._eventHandlerMap[eventId];
		}
	};
	// post event: eventId(number,string)
	dispathcer.post = function (eventId) {
		if ('string' != typeof(eventId) && 'number' != typeof(eventId)) {
			throw new Error("eventId is not number or string, it's type is " + typeof(eventId));
		}
		var handlers = this._eventHandlerMap[eventId];
		if (!(handlers instanceof Array) || 0 == handlers.length) {
			return false;
		}
		for (var i = 0; i < handlers.length; ++i) {
			if (handlers[i] && 'object' == typeof(handlers[i]) && 'function' == typeof(handlers[i].func)) {
				handlers[i].func.apply(handlers[i].target, Array.prototype.slice.call(arguments, 1));
			}
		}
		return true;
	};
	return dispathcer;
}
/**********************************************************************/
var EventCenter = (function () {
	var center = {};
    center._dispathcer = createEventDispatcher();
	// bind event
	center.bind = function(eventId, func, target, priority) {
		return this._dispathcer.subscribe(eventId, func, target, priority);
	};
	// unbind event
	center.unbind = function(eventId, func) {
		this._dispathcer.unsubscribe(eventId, func);
	};
	// post event
	center.post = function(eventId) {
		return this._dispathcer.post.apply(this._dispathcer, arguments);
	};
	return center;
})();
/**********************************************************************/